Add dispute resolution task

octaedro opened the pull request:
Fixes #1658

#### Changes proposed in this Pull Request

This PR adds a dispute resolution task for each dispute that the users have.

![screenshot-fermarichal jurassic tube-2021 06 11-10_56_43](https://user-images.githubusercontent.com/1314156/121697866-dbbcde00-caa3-11eb-8123-e8576ae61b10.png)


<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

- Create a new dispute. You can do it by buying one product with one of the [Stripe test disputes cards](https://stripe.com/docs/testing#disputes) (E.g.: `4000000000001976` use any date in the future and fill out `CVC` with any number).
- Verify the information displayed on the card is correct.
-  Click on the card. You should be redirected to the dispute's overview screen.
- Create another dispute and verify that the new dispute is also visible as a new dispute resolution task.
- Click on the new card and verify that is working as expected.

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

-------------------

### UPDATE

After merging [PR 2053](https://github.com/Automattic/woocommerce-payments/pull/2053), WCPay uses [@woocommerce/experimental 1.3.0](https://github.com/Automattic/woocommerce-payments/blob/develop/package.json#L101). A few modifications were added to the task list in that version that affects this PR and the [reconnect WooCommerce Payments](https://github.com/Automattic/woocommerce-payments/pull/1912) task.
The [issue 7199](https://github.com/woocommerce/woocommerce-admin/issues/7199) was created to fix it.

### UPDATE 2

Please review this PR after merging [PR 7263](https://github.com/woocommerce/woocommerce-admin/pull/7263), [PR 7300](https://github.com/woocommerce/woocommerce-admin/pull/7300), and having a new version of  `@woocommerce/experimental` released.

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

octaedro committed:
Add dispute resolution task

# Conflicts:
#	client/overview/index.js

# Conflicts:
#	client/overview/index.js
#	client/overview/task-list/tasks.js

octaedro committed:
Added tests

# Conflicts:
#	client/overview/task-list/test/tasks.js

octaedro committed:
Added time of day

octaedro committed:
Added changelog

# Conflicts:
#	readme.txt

# Conflicts:
#	readme.txt

# Conflicts:
#	readme.txt

octaedro committed:
Fixed comment for translators

octaedro committed:
Fixed completion task

mattsherman commented:
@octaedro This is marked as both "needs review" and "blocked", which seems like a conflict.

octaedro commented:
Hi @mattsherman,

> This is marked as both "needs review" and "blocked", which seems like a conflict.

Yes, you're right. Technically this PR is ready for review but it won't work as expected until a fix for the blocker issue is in place. The issue was added after this PR creation and should be fixed on `WooCommerce Admin`.
If you think that this PR should be labeled as `in progress` until the fix was pushed, I can change it 👍

mattsherman commented:
> If you think that this PR should be labeled as `in progress` until the fix was pushed, I can change it 👍

I think that is the most clear. If you want somebody to give a first-pass review/test of the PR with the outstanding issue, you can always ping somebody and ask them, as long as it is clear to them that it isn't ready for "final" review!

louwie17 reviewed (commented):
This looks good and tested well! I left one small suggestion in the code, and had a couple questions while testing it:
- Should we change the pagesize of the query to `5`? it currently is set to `25` which could make the things to do list very long. I wonder if we should show a different to do task item in that case? ( You have 20 disputed payments that needs your response )?
- Locally the disputes request takes a little while, so my todo list would render one item first, and after a little bit add the disputed items (on first load). This is not a huge deal, but I wonder if we can speed this up, maybe add caching on the back-end?
- The `Respond by` is not overly explicit (which is fine), it could be worth increasing the task level (maybe level 2 when there are 3 days left, and level 1 when there is one day left)?

These questions are not blockers, but I thought I would mention them here, incase we want to create issues for them.
cc: @LevinMedia @elizaan36 (for the above questions/thoughts)

octaedro committed:
Moved `getDisputesTasks` outside `getTasks

octaedro committed:
Updated @woocommerce/experimental version

octaedro committed:
Removed `eslint-disable camelcase`

octaedro committed:
Added package-lock.json

elizaan36 commented:
> Should we change the pagesize of the query to 5? it currently is set to 25 which could make the things to do list very long. I wonder if we should show a different to do task item in that case? ( You have 20 disputed payments that needs your response )?

@LevinMedia do we have any data showing the average/median number of disputes stores need to deal with on a regular basis, especially for larger stores? If there could be up to 25 in the list that could get problematic. One possibility is the footer accordion in use on the home screen task list that could be utilized here.

> Locally the disputes request takes a little while, so my todo list would render one item first, and after a little bit add the disputed items (on first load). This is not a huge deal, but I wonder if we can speed this up, maybe add caching on the back-end?

Anything we can do to make the tasks render at a consistent speed within the task list would be beneficial. Otherwise it looks messy and could convey a less reliable UI.

> The Respond by is not overly explicit (which is fine), it could be worth increasing the task level (maybe level 2 when there are 3 days left, and level 1 when there is one day left)?

My understanding is that disputes are not required tasks, and it's up to the merchant to take action on them. So changing the level based on time to respond seems overly complex.

mattsherman commented:
> These questions are not blockers, but I thought I would mention them here, incase we want to create issues for them.
> cc: @LevinMedia @elizaan36 (for the above questions/thoughts)

Great feedback. Those definitely sound like things that could be handled in follow-up issues/PRs.

octaedro commented on client/overview/task-list/tasks.js:
```diff
@@ -7,16 +7,70 @@ import { __, sprintf } from '@wordpress/i18n';
+		if ( ! disputes ) {
+			return [];
+		}
+		return disputes.map(
```
Good idea! I moved it in the commit 17b49d2

octaedro commented:
Thank you @louwie17 for your review,
I applied the suggested modifications and I also added the commit f5ed545 that modifies the @woocommerce/experimental version we're using here.
Could you give this PR another look?

louwie17 reviewed (approved):
Thanks for the updates @octaedro, this tested well, approving this!

> Anything we can do to make the tasks render at a consistent speed within the task list would be beneficial. Otherwise it looks messy and could convey a less reliable UI.

As Elizabeth mentioned, it would be worth addressing the loading issue (which I assume is not as bad on a live site versus a dev site), but that can be done in a separate PR.

octaedro committed:
Add ellipsis to dispute task (#2365)

* Added tests

# Conflicts:
#	client/overview/task-list/test/tasks.js

* Fixed completion task

# Conflicts:
#	client/overview/task-list/tasks.js

* Add dismiss option to task items

# Conflicts:
#	client/overview/task-list/tasks.js

* Add remind me later option to task items

# Conflicts:
#	client/overview/task-list/tasks.js

* Add delete option to task items

# Conflicts:
#	client/overview/task-list/tasks.js

* Refactor add ellipsis to dispute

* Fixed task list component

* Added animation to task-list item dismissal

* Added show prop to CollapsibleList

* Removed `eslint-disable camelcase`

* Changed animation to `custom`

* Added loading control for disputes

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>
# Conflicts:
#	includes/admin/class-wc-payments-admin.php

octaedro committed:
Added taskSort to task-list (#2328)

This commit adds the code to move the completed tasks to the bottom of the list.

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>
# Conflicts:
#	client/overview/task-list/test/tasks.js

LevinMedia commented:
@octaedro @louwie17  - after further review of the issue, @elizaan36 decided that we can simplify this task significantly. 

When I originally designed this, I hadn't fully considered the scope of the issue. Given that we're going to encounter merchant with both few disputes, and many disputes, We feel it's best to simply create a single task that says the number of disputes that need response, and point it to the disputes overview screen. 

I'll provide an updated design for this today.

LevinMedia commented:
With the new UX, the task list item should persist as long as there's a dispute that requires response. The number of disputes in the task will go up or down. 

<img width="498" alt="Screen Shot 2021-07-21 at 8 16 09 AM" src="https://user-images.githubusercontent.com/4500952/126514220-4e70ba3f-4485-4b36-8c92-e3db281f233c.png">

I'm curious if it would be possible to create individual task completed items for each dispute that either closes automatically, or has evidences submitted by a merchant. Each task completed item would link directly to the individual dispute that was either closed automatically, or had evidence submitted by the merchant.

<img width="506" alt="Screen Shot 2021-07-21 at 8 25 46 AM" src="https://user-images.githubusercontent.com/4500952/126515962-06be8635-9406-4a52-bc7e-a8c674a71f96.png">

mattsherman commented:
> @octaedro @louwie17 - after further review of the issue, @elizaan36 decided that we can simplify this task significantly.
> 
> When I originally designed this, I hadn't fully considered the scope of the issue. Given that we're going to encounter merchant with both few disputes, and many disputes, We feel it's best to simply create a single task that says the number of disputes that need response, and point it to the disputes overview screen.
> 
> I'll provide an updated design for this today.

@LevinMedia @elizaan36 @pmcpinto I do not disagree that this simplification is an improvement. Such a scope/requirements change will have an impact on the timeframe of completing this. We should do what we believe is the best experience, which sounds like what @LevinMedia proposes. Especially since we haven't shipped anything in regards to this functionality yet.

I'm just making sure that @pmcpinto recognizes that this change will result in the delivery of this taking longer. How much so, I am not sure -- once there is an updated design we can provide an estimate.

> I'm curious if it would be possible to create individual task completed items for each dispute that either closes automatically, or has evidences submitted by a merchant. Each task completed item would link directly to the individual dispute that was either closed automatically, or had evidence submitted by the merchant.

This is a departure from the current behavior of the task list to date, having completed task items appear that didn't originate from an individual uncompleted task item that previously. And, we would need to consider how to handle a large number of completed items in the task list as well.

It seems it would be better to just rely on the merchant using the disputes screen to view closed disputes.

pmcpinto commented:
Thanks for the heads up. Simplifying the UX seems to be the best approach and I agree with @mattsherman mentioned here:

> This is a departure from the current behavior of the task list to date, having completed task items appear that didn't originate from an individual uncompleted task item that previously. And, we would need to consider how to handle a large number of completed items in the task list as well.
> It seems it would be better to just rely on the merchant using the disputes screen to view closed disputes.

@elizaan36 @LevinMedia do we already have tracking in place to cover the usage of the features implemented in the WCPay dashboard? We should also include at least one event to track the usage of this task.

LevinMedia commented:
> This is a departure from the current behavior of the task list to date, having completed task items appear that didn't originate from an individual uncompleted task item that previously. And, we would need to consider how to handle a large number of completed items in the task list as well.

In that case, when the last dispute is completed, lets just check task as completed without changing any of the copy - the completed state would look like this: 

<img width="513" alt="Screen Shot 2021-07-22 at 11 23 08 AM" src="https://user-images.githubusercontent.com/4500952/126689676-7838ca3c-9567-46f5-8ab5-6f1afdb64694.png">

>We should also include at least one event to track the usage of this task.

@pmcpinto agree with you there. I'm looking into the status of additional event tracking.

mattsherman commented:
@LevinMedia @elizaan36 @pmcpinto From your standpoint is the re-design/re-scope of this complete from your point and ready for us to implement? Or, are there any further planned changes/reviews before we do so?

elizaan36 commented:
The update looks great to me, just one small tweak for the completed task style for consistency. The title changes to grey and helper text hidden on the completed task.

<img width="689" alt="Screen Shot 2021-07-27 at 12 46 38 PM" src="https://user-images.githubusercontent.com/5121465/127218298-833fda80-8bde-4b0f-8ed9-7680a2870e49.png">

@pmcpinto Good call on creating tracks events on the WCPay Overview. @LevinMedia let me know if I can assist with that.
